lib/dockerregistry/checks_test.go

```diff
@@ -317,3 +318,153 @@ func TestImageSizeCheck(t *testing.T) {
 			fmt.Sprintf("checkError: test: %v (ImageSizeCheck)\n", test.name))
 	}
 }
+
+// TestImageVulnCheck uses a fake populateRequests function and a fake
+// vulnerability producer. The fake vulnerability producer simply returns the
+// vulnerability occurrences that have been mapped to a given PromotionEdge in
+// order to simulate running the real check without having to make an api call
+// to the Container Analysis Service.
+func TestImageVulnCheck(t *testing.T) {
+	edge1 := reg.PromotionEdge{
+		SrcImageTag: reg.ImageTag{
+			ImageName: "foo",
+		},
+		Digest: "sha256:000",
+		DstImageTag: reg.ImageTag{
+			ImageName: "bar",
+		},
+	}
+	edge2 := reg.PromotionEdge{
+		SrcImageTag: reg.ImageTag{
+			ImageName: "bar",
+		},
+		Digest: "sha256:111",
+		DstImageTag: reg.ImageTag{
+			ImageName: "bar/1",
+		},
+	}
+	edge3 := reg.PromotionEdge{
+		SrcImageTag: reg.ImageTag{
+			ImageName: "bar",
+		},
+		Digest: "sha256:111",
+		DstImageTag: reg.ImageTag{
+			ImageName: "bar/2",
+		},
+	}
+
+	mkVulnProducerFake := func(
+		edgeVulnerabilities map[reg.Digest][]*grafeaspb.VulnerabilityOccurrence,
+	) reg.ImageVulnProducer {
+		return func(
+			edge reg.PromotionEdge,
+		) ([]*grafeaspb.VulnerabilityOccurrence, error) {
+			return edgeVulnerabilities[edge.Digest], nil
+		}
+	}
+
+	var tests = []struct {
+		name              string
+		severityThreshold int
+		edges             map[reg.PromotionEdge]interface{}
+		vulnerabilities   map[reg.Digest][]*grafeaspb.VulnerabilityOccurrence
+		expected          error
+	}{
+		{
+			"Severity under threshold",
+			int(grafeaspb.Severity_HIGH),
+			map[reg.PromotionEdge]interface{}{
+				edge1: nil,
+				edge2: nil,
+			},
+			map[reg.Digest][]*grafeaspb.VulnerabilityOccurrence{
+				"sha256:000": {
+					{
+						Severity: grafeaspb.Severity_LOW,
+					},
+				},
+				"sha256:111": {
+					{
+						Severity: grafeaspb.Severity_MEDIUM,
+					},
+				},
+			},
+			nil,
+		},
+		{
+			"Severity at threshold",
+			int(grafeaspb.Severity_HIGH),
+			map[reg.PromotionEdge]interface{}{
+				edge1: nil,
+				edge2: nil,
+			},
+			map[reg.Digest][]*grafeaspb.VulnerabilityOccurrence{
+				"sha256:000": {
+					{
+						Severity: grafeaspb.Severity_HIGH,
+					},
+				},
+				"sha256:111": {
+					{
+						Severity: grafeaspb.Severity_HIGH,
+					},
+				},
+			},
+			fmt.Errorf("VulnerabilityCheck: The following vulnerable images were found: bar (1), foo (1)"),
+		},
+		{
+			"Severity above threshold",
+			int(grafeaspb.Severity_MEDIUM),
+			map[reg.PromotionEdge]interface{}{
+				edge1: nil,
+				edge2: nil,
+			},
+			map[reg.Digest][]*grafeaspb.VulnerabilityOccurrence{
+				"sha256:000": {
+					{
+						Severity: grafeaspb.Severity_HIGH,
+					},
+				},
+				"sha256:111": {
+					{
+						Severity: grafeaspb.Severity_HIGH,
+					},
+					{
+						Severity: grafeaspb.Severity_CRITICAL,
+					},
+				},
+			},
+			fmt.Errorf("VulnerabilityCheck: The following vulnerable images were found: bar (2), foo (1)"),
```

Comment:
Can you make this error report each image and its digest (ideally, the FQIN)? And also show the vulnerability information as well (the entire thing we get from the API)?

In general, you never want to render a list of information on a single line. Imagine if there are 40 images with errors. You will force users to read a very long single line string. Not to mention, that would be a parsing nightmare if you wanted to use the output of this test to feed that into some automation downstream.

Reply:
The ImageVulnError type should be verbose enough (it contains all of the vulnerability data) and is the error that will be written to the Prow job logs and what the users will ultimately see. The error that is returned from Run() is just information I'm using for testing. I updated it to have the FQIN but I don't see the benefit of putting the entire vulnerability json return from Container Analysis into the error if I'm just using it for testing. The FQIN on the other hand is actually important for testing purposes.

Reply:
That's fine, I just want to make sure that we are actually showing all the information that we are dealing with at *some* point in the process. If it doesn't surface here in this particular case, but does get surfaced elsewhere, that works too.